### DISCUSSION TITLE
Templates for applications and packages

### DISCUSSION THREAD

**Discussion by ocots** [control-toolbox/control-toolbox.github.io#61] [General]
@jbcaillau I have started a [repo](https://github.com/control-toolbox/CTApp.jl) that I will use to make a template. For the moment, there will be one template for both applications and packages.

Once the repo is done:

- I will [create a template](https://docs.github.com/en/repositories/creating-and-managing-repositories/creating-a-template-repository).
- I think that the secrets have to be done for each new repo so I will add some instructions to complete setup when using the template.

What should be done after:

- change the name of the repo everywhere it appears:
  - `README.md`
  - `Project.toml`
  - `test/runtests.jl`
  - `test/test_default.jl`
  - `src/CTApp.jl`
  - `docs/Project.toml`
  - `docs/make.jl`
  - `docs/src/index.md`
  - `.github/ISSUE_TEMPLATE/config.yml`
- update UUID in `Project.toml` with `string(UUIDs.uuid4())`
- update authors in `Project.toml`
- add documenter keys with [DocumenterTools.genkeys](https://documenter.juliadocs.org/stable/lib/public/#DocumenterTools.genkeys). Don't forget to allow write access for the deploy key. The function which generates the keys may be found [here](https://github.com/JuliaDocs/DocumenterTools.jl/blob/master/src/genkeys.jl).
- add codecov secret for coverage: go to [codecov/control-toolbox](https://app.codecov.io/gh/control-toolbox) and click on Configure and then copy the `CODECOV_TOKEN` as a new repository secret in configuration > secrets and variable > actions.
- update `assignees` for new issues, cf `.github/ISSUE_TEMPLATE` 
- active Discussions in Settings.
- create a first release. See [here](https://github.com/control-toolbox/preconditioning/issues/8) for new release. 
- check that github pages are deployed from branch `gh-pages`:

<img width="1129" alt="Capture d’écran 2024-08-28 à 22 40 01" src="https://github.com/user-attachments/assets/4bdfe444-f178-4194-9ecd-0602dad28f19">
- Give permissions to github actions:

![Capture d’écran 2024-08-29 à 09 24 59](https://github.com/user-attachments/assets/6589dfac-b2dd-44ee-a875-04b7cd562311)

**Comment by ocots** [comment]
The name appears many times. It should be better to automate this.

**Comment by ocots** [comment]
Maybe we could make an action put inside the repo that automate all the remaining stuff to do.

**Reply by jbcaillau** [reply]
very nice, thanks @ocots may be this should be moved to some wiki (the one in the associated repo).

**Reply by ocots** [reply]
Yes I wanted to test before. So now it seems to work. I can make a wiki page.

**Comment by ocots** [comment]
Maybe it is better to add an `INSTRUCTIONS.md` file in CTApp.jl instead of a wiki page.

**Reply by jbcaillau** [reply]
please choose the most standard way 🤞🏾

**Comment by ocots** [comment]
En fait, vu qu'une appli et un package partage le même template et qu'une appli peut devenir un package, je propose que même pour une appli, on donne un nom du genre `MyApp.jl`. Ca simplifie aussi la mise en place.
